File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/drive/MecanumDriveSubsystem.java
package org.firstinspires.ftc.teamcode.subsystems.drive;

import com.acmerobotics.roadrunner.Action;
import com.acmerobotics.roadrunner.ftc.Actions;
import com.acmerobotics.roadrunner.Pose2d;
import com.acmerobotics.roadrunner.PoseVelocity2d;
import com.acmerobotics.roadrunner.TrajectoryActionBuilder;
import com.acmerobotics.roadrunner.Vector2d;
import com.arcrobotics.ftclib.command.SubsystemBase;
import com.qualcomm.robotcore.hardware.HardwareMap;

import java.util.List;

public class MecanumDriveSubsystem extends SubsystemBase {

    public final MecanumDrive drive;
    private final boolean fieldCentric;

    public MecanumDriveSubsystem(HardwareMap hardwareMap, Pose2d startPos, boolean isFieldCentric) {
        this.drive = new MecanumDrive(hardwareMap, startPos);
        fieldCentric = isFieldCentric;

        drive.actionBuilder(new Pose2d(0, 0, 0)).getDispResolution();
    }

    public void setPoseEstimate(Pose2d pose) {
        drive.pose = pose;
    }

    public PoseVelocity2d updatePoseEstimate() {
        return drive.updatePoseEstimate();
    }

    public void drive(double leftY, double leftX, double rightX) {
        Pose2d poseEstimate = getPoseEstimate();

        Vector2d input = new Vector2d(-leftY, -leftX);

        drive.setDrivePowers(
                new PoseVelocity2d(
                        input,
                        -rightX
                )
        );
    }

    public void setDrivePower(PoseVelocity2d drivePower) {
        drive.setDrivePowers(drivePower);
    }

    public Pose2d getPoseEstimate() {
        return drive.pose;
    }

    public TrajectoryActionBuilder trajectoryActionBuilder(Pose2d startPose) {
        return drive.actionBuilder(startPose);
    }

    public void followAction(Action action) {
        Actions.runBlocking(action);
    }

    public boolean isBusy() {
        return drive.leftBack.isBusy();
    }

    public void turn(double radians) {
        Actions.runBlocking(drive.actionBuilder(getPoseEstimate())
                .turn(radians)
                .build());
    }

    public void stop() {
        drive(0, 0, 0);
    }

    public Localizer getLocalizer() {
        return drive.localizer;
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/drive/MecanumDriveSubsystem.kt
package org.firstinspires.ftc.teamcode.subsystems.drive

import com.acmerobotics.roadrunner.Pose2d
import com.acmerobotics.roadrunner.PoseVelocity2d
import com.acmerobotics.roadrunner.Vector2d
import com.arcrobotics.ftclib.command.SubsystemBase

/**
 * A subsystem that uses the [MecanumDrive] class.
 * This periodically calls [MecanumDrive.update] which runs the internal
 * state machine for the mecanum drive. All movement/following is async to fit the paradigm.
 */
class MecanumDriveSubsystem(drive: MecanumDrive, isFieldCentric: Boolean):
    SubsystemBase() {
    private val drive: MecanumDrive
    private val fieldCentric: Boolean

    init {
        this.drive = drive
        fieldCentric = isFieldCentric
    }

    fun updatePoseEstimate() {
        drive.updatePoseEstimate()
    }

    fun drive(leftY: Double, leftX: Double, rightX: Double) {
        val poseEstimate = poseEstimate
        val input: Vector2d = Vector2d(-leftY, -leftX)
        drive.setDrivePowers(
            PoseVelocity2d(
                Vector2d(
                    -leftY,
                    -leftX
                ),
                -rightX
            )
        )
    }

    fun setDrivePower(drivePower: PoseVelocity2d?) {
        drive.setDrivePowers(drivePower)
    }

    fun followAction(action: MecanumDrive.FollowTrajectoryAction?) {

    }

    var poseEstimate: Pose2d
        get() = drive.pose
        set(pose) {
            drive.pose = pose
        }

    fun stop() {
        drive(0.0, 0.0, 0.0)
    }
}